### PR TITLE
introduce uORB::PublicationQueued and transition most orb_advertise_queue usage

### PR DESCRIPTION
**Pull request by dagar**

### CODE REVIEW

**Comment by dagar** [comment]
@mcsauder interested in this one?

**Comment by mcsauder** [comment]
@dagar , I'll get it reviewed and test it when I get back from travels around Aug 1.

File: src/drivers/drv_led.h [hunk 1]
```diff
@@ -49,14 +49,14 @@
 #define BOARD_MAX_LEDS 4
 #endif
 
+#if BOARD_MAX_LEDS < LED_CONTROL_ORB_QUEUE_LENGTH
```
Comment (julianoes) [inline src/drivers/drv_led.h]: This checks looks like it's the wrong way round. And I can't find `LED_CONTROL_ORB_QUEUE_LENGTH` anywhere in this diff. Did you mean `led_control_s::ORB_QUEUE_LENGTH`?

File: src/drivers/gps/gps.cpp [hunk 2]
```diff
@@ -172,9 +173,10 @@ class GPS : public ModuleBase<GPS>
 	const Instance 			_instance;
 
 	uORB::Subscription		_orb_inject_data_sub{ORB_ID(gps_inject_data)};
-	orb_advert_t			_dump_communication_pub{nullptr};		///< if non-null, dump communication
+	uORB::PublicationQueued<gps_dump_s>	_dump_communication_pub{ORB_ID(gps_dump)};
 	gps_dump_s			*_dump_to_device{nullptr};
 	gps_dump_s			*_dump_from_device{nullptr};
+	bool				_dump_communication{false};			///< if true, dump communication
```
Comment (julianoes) [inline src/drivers/gps/gps.cpp]: I find the comment is not necessary. If it's not clear enough, what about calling it `_should_dump_communication`?

File: src/lib/systemlib/mavlink_log.cpp [hunk 3]
```diff
@@ -59,31 +58,23 @@ __EXPORT void mavlink_vasprintf(int severity, orb_advert_t *mavlink_log_pub, con
...
 
-	if (*mavlink_log_pub != NULL) {
+	if (*mavlink_log_pub != nullptr) {
 		orb_publish(ORB_ID(mavlink_log), *mavlink_log_pub, &log_msg);
 
 	} else {
-		*mavlink_log_pub = orb_advertise_queue(ORB_ID(mavlink_log),
-						       &log_msg,
-						       MAVLINK_LOG_QUEUE_SIZE);
+		*mavlink_log_pub = orb_advertise_queue(ORB_ID(mavlink_log), &log_msg, MAVLINK_LOG_QUEUE_SIZE);
```
Comment (julianoes) [inline src/lib/systemlib/mavlink_log.cpp]: You didn't want to change this one to `PublicationQueued`?

File: src/modules/uORB/PublicationQueued.hpp [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+	 */
+	bool publish(const T &data)
+	{
+		bool updated = false;
+
+		if (_handle != nullptr) {
+			updated = (orb_publish(_meta, _handle, &data) == PX4_OK);
+
+		} else {
+			orb_advert_t handle = orb_advertise_queue(_meta, &data, T::ORB_QUEUE_LENGTH);
```
Comment (julianoes) [inline src/modules/uORB/PublicationQueued.hpp]: Nice! No more passing of this queue length, that's good.

File: src/lib/systemlib/mavlink_log.cpp (hunk 3 above)
Comment (dagar) [reply src/lib/systemlib/mavlink_log.cpp]: Mavlink log (mavlink_log_info, mavlink_log_warning, etc) is kind of it's own thing that needs a bit more thought.

**Comment by mcsauder** [comment]
This PR flies great, flight tested on pixhawk 4 mini, 250sized quad: https://review.px4.io/plot_app?log=114ba234-cbc6-4cc3-afd9-a9cf5c3393b2

File: src/drivers/gps/gps.cpp (hunk 2 above)
Comment (dagar) [reply src/drivers/gps/gps.cpp]: Updated.

File: src/drivers/drv_led.h (hunk 1 above)
Comment (dagar) [reply src/drivers/drv_led.h]: Fixed.

Comment (dagar) [reply src/drivers/drv_led.h]: The ORB queue length must be >= the number of lights on a board. It's cleared in the updated commit.

File: src/lib/systemlib/mavlink_log.cpp (hunk 3 above)
Comment (julianoes) [reply src/lib/systemlib/mavlink_log.cpp]: Ok.

**Pull request merged**